File: FSharp/PowerOfTwo.fs
module PowerOfTwo

open System
open System.IO

let logPath = Path.Combine(__SOURCE_DIRECTORY__, "PowerOfTwo.log")
let logFile = new StreamWriter(logPath, false)

let header = "Computing powers of two (Big Integer)"
let separator = "======================================"

logFile.WriteLine(header)
logFile.WriteLine(separator)
logFile.WriteLine("")
logFile.Flush()

printfn "%s" header
printfn "%s" separator
printfn ""

let rec computePowers (value: bigint) (power: int) =
    let result = sprintf "2^%d = %A" power value
    logFile.WriteLine(result)
    logFile.Flush()
    printfn "%s" result
    computePowers (value * 2I) (power + 1)

try
    computePowers 1I 0
finally
    logFile.Close()

File: Dart/power_of_two.dart
import 'dart:io';

void main() {
  final logPath = '${Directory.current.path}${Platform.pathSeparator}PowerOfTwo.log';
  final logFile = File(logPath);

  final header = 'Computing powers of two (Big Integer)';
  final separator = '======================================';

  // Write to log file
  logFile.writeAsStringSync('$header\n$separator\n\n');

  // Print to console
  print(header);
  print(separator);
  print('');

  BigInt value = BigInt.one;
  int power = 0;

  try {
    while (power < 1000) {
      final result = '2^$power = $value';

      // Write to log file
      logFile.writeAsStringSync('$result\n', mode: FileMode.append);

      // Print to console
      print(result);

      value = value * BigInt.two;
      power++;
    }
  } catch (e) {
    print('Error: $e');
  }
}
